### PR TITLE
Story/vspc 23

### PR DESCRIPTION
**Pull request by renilJoseph**
cleared merge conflicts. Please verify.

### CODE REVIEW

File: vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java [hunk 1]
```diff
@@ -183,4 +221,34 @@ public IExternalLinkDisplay createExternalLink(String title, ISpace source, floa
...
+        IModule target = moduleManager.getModule(linkedModuleId);
+        IModuleLink link = moduleLinkFactory.createModuleLink(title, source);
+        link.setModule(target);
+        moduleLinkRepo.save((ModuleLink) link);
+
+        IModuleLinkDisplay display = moduleLinkDisplayFactory.createModuleLinkDisplay(link);
+        display.setPositionX(positionX);
+        display.setPositionY(positionY);
+        display.setRotation(rotation);
+        display.setType(displayType != null ? displayType : DisplayType.ARROW);
```
Comment (jdamerow) [inline vspace/src/main/java/edu/asu/diging/vspace/core/services/impl/LinkManager.java]: let's create a new display type module that is the default for module links

File: vspace/src/main/webapp/WEB-INF/views/exhibition/space.jsp [hunk 2]
```diff
@@ -42,6 +42,34 @@ function drawLinks() {
...
+        var posY = $("#space").position().top;
+        var link = $('<a></a>');
+        link.attr('href', '<c:url value="/exhibit/module/${link.link.module.id}" />');
+        
+        if ("${link.type}" == 'ALERT') {
+            var linkDisplay = $('<div class="alert alert-primary" role="alert">');
+        } else if ("${link.type}" == 'IMAGE' && "${link.image}" != '') {
+            var linkDisplay = $('<img id="${link.image.id}" src="<c:url value="/api/image/${link.image.id}" />" />');
+        } else {
+            var linkDisplay = $('<span data-feather="navigation-2" class="flex"></span>');
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/exhibition/space.jsp]: let's use the font awesome open book icon by default for modules (fas fa-book-open)

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp [hunk 3]
```diff
@@ -46,6 +47,42 @@ $( document ).ready(function() {
...
+	{
+		var link;
+		var posX = $("#bgImage").position().left;
+		var posY = $("#bgImage").position().top;
+		if ("${link.type}" == "ALERT") {
+			link = $('<div class="alert alert-primary" role="alert" data-link-id="${link.link.id}"><p>${link.link.name}</p></div>');
+		} else if ("${link.type}" == "IMAGE" && "${link.image.id}" != "") {
+           link = $('<img id="${link.image.id}" data-link-id="${link.link.id}" src="<c:url value="/api/image/${link.image.id}" />" />');
+		}  else {
+			link = $('<span data-link-id="${link.link.id}"><span data-feather="navigation-2" class="flex"></span></span><p class="mlabel-${loop.index}" data-link-id="${link.link.id}">${link.link.name}</p>');
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp]: same as on exhibition page, let's use the open book icon

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp [hunk 4]
```diff
@@ -187,6 +245,54 @@ $( document ).ready(function() {
...
+		e.preventDefault();
+		
+		if (storeX == undefined || storeY == undefined) {
+			$("#errorMsg").text("Please click on the image to specify where the new link should be located.")
+			$('#errorAlert').show();
+			return;
+		}
+		
+		var linkedModules = $("#linkedModule").val();
+		console.log(linkedModules);
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp]: remove

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp [hunk 5]
```diff
@@ -187,6 +245,54 @@ $( document ).ready(function() {
...
+	        	var linkData = JSON.parse(data);
+	        	$("#bgImage").off("click");
+	        	moduleLinkInfo["id"] = linkData["id"];
+	            showModuleLink(moduleLinkInfo, true);
+	            $("#module_label").attr("id","");
+	            $("#link").attr("id","");
+	            $("#createModuleLinkAlert").hide();  
+	            $("#errorMsg").text("")
+	            $('#errorAlert').hide();
+	        }
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp]: indentation

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp [hunk 6]
```diff
@@ -304,6 +435,59 @@ $( document ).ready(function() {
...
+			link.attr("data-link-id", moduleLink["id"]);
+			link.css('cursor', 'pointer');
+			link.click(function(e) {
+				makeModuleLinksEditable(moduleLink["moduleLinkLabel"], moduleLink["id"]);
+	        });
+			module_label.attr("data-link-id", moduleLink["id"]);
+			module_label.css('cursor', 'pointer');
+			module_label.click(function(e) {
+                makeModuleLinksEditable(moduleLink["moduleLinkLabel"], moduleLink["id"]);
+            });
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp]: indentation

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp [hunk 7]
```diff
@@ -422,127 +645,227 @@ ${space.description}
...
+
+		<div class="row">
+			<div class="col-sm-4">
+				<label><small>Type:</small> </label>
+			</div>
+			<div class="col-sm-8">
+				<select id="type" name="type" class="form-control-xs target">
+					<option selected value="">Choose...</option>
+					<option value="IMAGE">Image</option>
+					<option value="ARROW">Link</option>
```
Comment (jdamerow) [inline vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp]: as mentioned above, let's make that type module.

**Comment by jdamerow** [comment]
please reopen the previous pull request after you have made changes

File: vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp (hunk 5 above)
Comment (renilJoseph) [reply vspace/src/main/webapp/WEB-INF/views/staff/spaces/space.jsp]: could you please clarify whats wrong here. In my eclipse indentation seems fine.

**Pull request merged**